tests/integration-tests/tests/cloudwatch_logging/test_cloudwatch_logging.py

```diff
@@ -0,0 +1,460 @@
+        outputs = {}
+        result_files = self._run_command_on_master("ls {}".format(out_dir))
+        for hostname in result_files.split():
+            path = "/".join((out_dir, hostname))
+            outputs[hostname] = self._run_command_on_master("sudo cat {}".format(path))
+        self._run_command_on_master("rm -rf {}".format(out_dir))
+        return outputs
+
+    def _populate_master_log_existence(self):
+        """Figure out which of the relevant logs for the MasterServer don't exist."""
+        for log_path, log_dict in self._cluster_log_state["MasterServer"]["logs"].items():
+            cmd = "[[ -n `ls {path}` ]] && echo exists || echo does not exist".format(path=log_path)
+            output = self._run_command_on_master(cmd)
+            log_dict["exists"] = output == "exists"
+
+    def _populate_compute_log_existence(self):
+        """Figure out which of the relevant logs for the ComputeFleet nodes don't exist."""
+        for log_dict in self._relevant_logs["ComputeFleet"]:
+            cmd = "[[ -n `ls {path}` ]] && echo {{redirect}} exists || " "echo {{redirect}} does not exist".format(
+                path=log_dict["file_path"]
+            )
+            hostname_to_output = self._run_command_on_computes(cmd)
+            for hostname, output in hostname_to_output.items():
+                node_log_dict = self._cluster_log_state["ComputeFleet"][hostname]["logs"][log_dict["file_path"]]
+                node_log_dict["exists"] = output == "exists"
+
+    def _populate_log_existence(self):
+        """Figure out which of the relevant logs for each node type don't exist."""
+        self._populate_master_log_existence()
+        self._populate_compute_log_existence()
+        LOGGER.debug("After populating log existence:\n{}".format(self._dump_cluster_log_state()))
+
+    def _populate_master_log_emptiness_and_tail(self):
+        """Figure out which of the relevant logs for the MasterServer are empty."""
+        for log_path, log_dict in self._cluster_log_state["MasterServer"]["logs"].items():
+            if not log_dict["exists"]:
+                continue
+            output = self._run_command_on_master("sudo tail -n 1 {path}".format(path=log_path))
+            log_dict["is_empty"] = output == ""
+            log_dict["tail"] = output
+
+    def _populate_compute_log_emptiness_and_tail(self):
+        """Figure out which of the relevant logs for the ComputeFleet nodes are empty."""
+        for log_dict in self._relevant_logs["ComputeFleet"]:
+            # If this file doesn't exist on any of the computes, don't assert success
+            assert_success = True
+            for _, compute_dict in self._cluster_log_state["ComputeFleet"].items():
+                if not compute_dict["logs"][log_dict["file_path"]]["exists"]:
+                    assert_success = False
+                    break
+            cmd = "sudo tail {{redirect}} -n 1 {path}".format(path=log_dict["file_path"])
```

Comment:
Where does `{{redirect}}` come from?

Reply:
It gets `format`'ed in by the `_run_command_on_computes`. I did it this way because I couldn't figure out a better way for commands that required piping (like the one in `_populate_compute_log_existence`).